torch/_refs/__init__.py

```diff
@@ -932,3 +951,123 @@ def sum(
         dtype=dtype,
         output_dtype_kind=REDUCTION_OUTPUT_TYPE_KIND.SAME,
     )
+
+
+def cat(
+    tensors: TensorSequenceType, dim: int = 0, out: TensorLikeType = None
+) -> TensorLikeType:
+    if len(tensors) == 0:
+        msg = "cat expects at least one tensor, but received zero!"
+        raise ValueError(msg)
+
+    _dim = utils.canonicalize_dims(tensors[0].ndim, dim)
+    dtype, _ = _elementwise_dtypes(
+        *tensors, type_promotion=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
+    )
+
+    _tensors = _convert_dtype(*tensors, dtype=dtype)
+    result = prims.concatenate(_tensors, _dim)
+
+    if out is not None:
+        out = _maybe_resize_out(out, result.shape)
+        return copy_to(out, result, allow_cross_device=False)  # type: ignore[arg-type]
```

Comment:
Isn't it weird that all prims are not type-promoting except copy_to that is? Would it be better to instead check `out` dtype here, and not rely on `copy_to` to error out? If refs are documentation in code, then making error conditions clear is one of the goals.

Reply:
Let's reconsider this when we apply out handling as a wrapper?